File: apps/studio-web/src/features/animation-editor/animationCurves.ts
import type { AuthoredAnimationTrackV1 } from "../animation-studio/types";

export interface AnimationEditorCurveTrackV1 {
  readonly schemaVersion: 1;
  readonly id: string;
  readonly targetNodeId: number;
  readonly path: "TRANSLATION" | "ROTATION";
  readonly keys: readonly {
    readonly id: string;
    readonly timeSeconds: number;
    readonly value: readonly number[];
    readonly inTangent: readonly number[] | null;
    readonly outTangent: readonly number[] | null;
  }[];
}

export interface AnimationCurveResamplePolicyV1 {
  readonly schemaVersion: 1;
  readonly maxPositionError: number;
  readonly maxAngularErrorRadians: number;
  readonly maxSubdivisionDepth: number;
}

export interface AnimationCurveResampleReportV1 {
  readonly schemaVersion: 1;
  readonly editorKeyCount: number;
  readonly linearKeyCount: number;
  readonly maxPositionError: number;
  readonly maxAngularErrorRadians: number;
  readonly track: AuthoredAnimationTrackV1;
  readonly fingerprintSha256: string;
}

export function createAutoCurveFromTrackV1(
  track: AuthoredAnimationTrackV1,
): AnimationEditorCurveTrackV1 {
  return {
    schemaVersion: 1,
    id: track.id,
    targetNodeId: track.targetNodeId,
    path: track.path,
    keys: track.keyframes.map((key, index, keys) => {
      if (track.path === "ROTATION") {
        return {
          id: key.id,
          timeSeconds: key.timeSeconds,
          value: key.value,
          inTangent: null,
          outTangent: null,
        };
      }
      const previous = keys[Math.max(0, index - 1)]!;
      const next = keys[Math.min(keys.length - 1, index + 1)]!;
      const duration = next.timeSeconds - previous.timeSeconds;
      const tangent = duration <= 1e-8
        ? [0, 0, 0]
        : [0, 1, 2].map((axis) => (
            ((next.value[axis] ?? 0) - (previous.value[axis] ?? 0)) / duration
          ));
      return {
        id: key.id,
        timeSeconds: key.timeSeconds,
        value: key.value,
        inTangent: tangent,
        outTangent: tangent,
      };
    }),
  };
}

export function parseAnimationCurveResampleReportV1(
  json: string,
): AnimationCurveResampleReportV1 {
  const value = JSON.parse(json) as Partial<AnimationCurveResampleReportV1>;
  if (
    value.schemaVersion !== 1
    || !Number.isSafeInteger(value.editorKeyCount)
    || !Number.isSafeInteger(value.linearKeyCount)
    || !Number.isFinite(value.maxPositionError)
    || !Number.isFinite(value.maxAngularErrorRadians)
    || value.track === null
    || typeof value.track !== "object"
    || typeof value.fingerprintSha256 !== "string"
    || !/^[0-9a-f]{64}$/iu.test(value.fingerprintSha256)
  ) {
    throw new Error("Core returned an invalid curve-resample report.");
  }
  return value as AnimationCurveResampleReportV1;
}

File: crates/m2a-core/tests/animation_curves.rs
use m2a_core::{
    animation_curves::{
        AnimationCurveKeyV1, AnimationCurveResamplePolicyV1, AnimationEditorCurveTrackV1,
        resample_animation_curve_to_linear_v1,
    },
    animation_studio::AuthoredAnimationTrackPathV1,
};

#[test]
fn cubic_translation_is_adaptively_resampled_to_linear_keys() {
    let curve = AnimationEditorCurveTrackV1 {
        schema_version: 1,
        id: "hand-translation".into(),
        target_node_id: 7,
        path: AuthoredAnimationTrackPathV1::Translation,
        keys: vec![
            AnimationCurveKeyV1 {
                id: "start".into(),
                time_seconds: 0.0,
                value: vec![0.0, 0.0, 0.0],
                in_tangent: Some(vec![0.0, 0.0, 0.0]),
                out_tangent: Some(vec![4.0, 0.0, 0.0]),
            },
            AnimationCurveKeyV1 {
                id: "end".into(),
                time_seconds: 1.0,
                value: vec![1.0, 0.0, 0.0],
                in_tangent: Some(vec![-3.0, 0.0, 0.0]),
                out_tangent: Some(vec![0.0, 0.0, 0.0]),
            },
        ],
    };
    let report = resample_animation_curve_to_linear_v1(
        &curve,
        &AnimationCurveResamplePolicyV1 {
            schema_version: 1,
            max_position_error: 0.01,
            max_angular_error_radians: 0.01,
            max_subdivision_depth: 10,
        },
    )
    .unwrap();

    assert!(report.linear_key_count > 2);
    assert_eq!(
        report.track.interpolation,
        m2a_core::mdl::MdlAnimationInterpolationV1::Linear
    );
    assert_eq!(report.track.keyframes.first().unwrap().time_seconds, 0.0);
    assert_eq!(report.track.keyframes.last().unwrap().time_seconds, 1.0);
    assert_eq!(report.fingerprint_sha256.len(), 64);
}

#[test]
fn rotation_curve_rejects_unsafe_component_tangents() {
    let curve = AnimationEditorCurveTrackV1 {
        schema_version: 1,
        id: "hand-rotation".into(),
        target_node_id: 7,
        path: AuthoredAnimationTrackPathV1::Rotation,
        keys: vec![
            AnimationCurveKeyV1 {
                id: "a".into(),
                time_seconds: 0.0,
                value: vec![0.0, 0.0, 0.0, 1.0],
                in_tangent: None,
                out_tangent: Some(vec![0.0; 4]),
            },
            AnimationCurveKeyV1 {
                id: "b".into(),
                time_seconds: 1.0,
                value: vec![0.0, 0.0, 0.0, -1.0],
                in_tangent: None,
                out_tangent: None,
            },
        ],
    };
    assert!(
        resample_animation_curve_to_linear_v1(
            &curve,
            &AnimationCurveResamplePolicyV1 {
                schema_version: 1,
                max_position_error: 0.01,
                max_angular_error_radians: 0.01,
                max_subdivision_depth: 8,
            }
        )
        .is_err()
    );
}
